File: backend/internal/services/recipe_collection_service.go
package services

import (
	"errors"
	"fmt"

	"github.com/yourusername/scentora-backend/internal/models"
	"github.com/yourusername/scentora-backend/internal/repository"
)

type RecipeCollectionService struct {
	collectionRepo *repository.RecipeCollectionRepository
	recipeRepo     *repository.RecipeRepository
}

func NewRecipeCollectionService(
	collectionRepo *repository.RecipeCollectionRepository,
	recipeRepo *repository.RecipeRepository,
) *RecipeCollectionService {
	return &RecipeCollectionService{
		collectionRepo: collectionRepo,
		recipeRepo:     recipeRepo,
	}
}

// CreateCollection creates a new collection
func (s *RecipeCollectionService) CreateCollection(userID string, req *models.CreateRecipeCollectionRequest) (*models.RecipeCollection, error) {
	// Validate input
	if req.Name == "" {
		return nil, errors.New("collection name is required")
	}

	// Check for duplicate name
	exists, err := s.collectionRepo.Exists(userID, req.Name)
	if err != nil {
		return nil, fmt.Errorf("failed to check collection existence: %w", err)
	}
	if exists {
		return nil, errors.New("collection with this name already exists")
	}

	// Create collection
	collection := &models.RecipeCollection{
		UserID:      userID,
		Name:        req.Name,
		Description: req.Description,
	}
	err = s.collectionRepo.Create(collection)
	if err != nil {
		return nil, fmt.Errorf("failed to create collection: %w", err)
	}

	return collection, nil
}

// GetCollection retrieves a collection by ID
func (s *RecipeCollectionService) GetCollection(collectionID, userID string) (*models.RecipeCollection, error) {
	collection, err := s.collectionRepo.FindByID(collectionID, userID)
	if err != nil {
		return nil, fmt.Errorf("collection not found: %w", err)
	}

	// Get recipe count
	if err != nil {
		return nil, fmt.Errorf("failed to count recipes: %w", err)
	}

	return collection, nil
}

// ListCollections retrieves all collections for a user
func (s *RecipeCollectionService) ListCollections(userID string, limit, offset int) ([]*models.RecipeCollection, error) {
	collections, err := s.collectionRepo.FindAll(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to list collections: %w", err)
	}

	return collections, nil
}

// UpdateCollection updates a collection
func (s *RecipeCollectionService) UpdateCollection(collectionID, userID string, req *models.UpdateRecipeCollectionRequest) (*models.RecipeCollection, error) {
	// Get existing collection
	collection, err := s.collectionRepo.FindByID(collectionID, userID)
	if err != nil {
		return nil, fmt.Errorf("collection not found: %w", err)
	}

	// Validate name change
	if req.Name != nil && *req.Name != collection.Name {
		if *req.Name == "" {
			return nil, errors.New("collection name is required")
		}

		exists, err := s.collectionRepo.Exists(userID, *req.Name)
		if err != nil {
			return nil, fmt.Errorf("failed to check collection existence: %w", err)
		}
		if exists {
			return nil, errors.New("collection with this name already exists")
		}
		collection.Name = *req.Name
	}

	// Update description
	if req.Description != nil {
		collection.Description = req.Description
	}

	// Save changes
	err = s.collectionRepo.Update(collection)
	if err != nil {
		return nil, fmt.Errorf("failed to update collection: %w", err)
	}

	return collection, nil
}

// DeleteCollection deletes a collection
func (s *RecipeCollectionService) DeleteCollection(collectionID, userID string) error {
	// Verify collection exists and belongs to user
	_, err := s.collectionRepo.FindByID(collectionID, userID)
	if err != nil {
		return fmt.Errorf("collection not found: %w", err)
	}

	// Delete collection (cascade will handle membership records)
	err = s.collectionRepo.Delete(userID, collectionID)
	if err != nil {
		return fmt.Errorf("failed to delete collection: %w", err)
	}

	return nil
}

// AddRecipeToCollection adds a recipe to a collection
func (s *RecipeCollectionService) AddRecipeToCollection(collectionID, recipeID, userID string) error {
	// Verify collection exists and belongs to user
	_, err := s.collectionRepo.FindByID(collectionID, userID)
	if err != nil {
		return fmt.Errorf("collection not found: %w", err)
	}

	// Verify recipe exists and belongs to user
	_, err = s.recipeRepo.FindByID(userID, recipeID)
	if err != nil {
		return fmt.Errorf("recipe not found: %w", err)
	}

	// Check if already in collection
	exists, err := s.collectionRepo.IsRecipeInCollection(collectionID, recipeID)
	if err != nil {
		return fmt.Errorf("failed to check recipe membership: %w", err)
	}
	if exists {
		// Already in collection, no error
		return nil
	}

	// Add recipe to collection
	err = s.collectionRepo.AddRecipe(&models.RecipeCollectionMember{CollectionID: collectionID, RecipeID: recipeID})
	if err != nil {
		return fmt.Errorf("failed to add recipe to collection: %w", err)
	}

	return nil
}

// RemoveRecipeFromCollection removes a recipe from a collection
func (s *RecipeCollectionService) RemoveRecipeFromCollection(collectionID, recipeID, userID string) error {
	// Verify collection exists and belongs to user
	_, err := s.collectionRepo.FindByID(collectionID, userID)
	if err != nil {
		return fmt.Errorf("collection not found: %w", err)
	}

	// Verify recipe exists and belongs to user
	_, err = s.recipeRepo.FindByID(userID, recipeID)
	if err != nil {
		return fmt.Errorf("recipe not found: %w", err)
	}

	// Remove recipe from collection
	err = s.collectionRepo.RemoveRecipe(collectionID, recipeID)
	if err != nil {
		return fmt.Errorf("failed to remove recipe from collection: %w", err)
	}

	return nil
}

// GetRecipeCollections retrieves all collections containing a recipe
func (s *RecipeCollectionService) GetRecipeCollections(recipeID, userID string) ([]*models.RecipeCollection, error) {
	// Verify recipe exists and belongs to user
	_, err := s.recipeRepo.FindByID(userID, recipeID)
	if err != nil {
		return nil, fmt.Errorf("recipe not found: %w", err)
	}

	collections, err := s.collectionRepo.GetCollectionsByRecipeID(recipeID, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get collections: %w", err)
	}

	return collections, nil
}

File: backend/src/services/recipeCollection.service.ts
import { prisma } from '../prisma.js';
import { NotFoundError, ValidationError } from '../utils/errors.js';

function collectionToApi(collection: any) {
  return {
    _id: collection.id,
    userId: collection.userId,
    name: collection.name,
    ...(collection.description != null ? { description: collection.description } : {}),
    createdAt: collection.createdAt,
    updatedAt: collection.updatedAt,
  };
}

export async function createCollection(userId: string, data: { name: string; description?: string }) {
  const exists = await prisma.recipeCollection.findFirst({ where: { userId, name: data.name } });
  if (exists) throw new ValidationError('collection with this name already exists');

  const collection = await prisma.recipeCollection.create({
    data: {
      userId,
      name: data.name,
      description: data.description ?? null,
    },
  });

  return collectionToApi(collection);
}

export async function listCollections(userId: string, limit = 100, offset = 0) {
  const collections = await prisma.recipeCollection.findMany({
    where: { userId },
    orderBy: { createdAt: 'desc' },
    take: limit,
    skip: offset,
  });
  return collections.map(collectionToApi);
}

export async function getCollection(collectionId: string, userId: string) {
  const collection = await prisma.recipeCollection.findFirst({
    where: { id: collectionId, userId },
  });
  if (!collection) throw new NotFoundError('collection not found');
  return collectionToApi(collection);
}

export async function updateCollection(collectionId: string, userId: string, data: {
  name?: string;
  description?: string | null;
}) {
  const collection = await prisma.recipeCollection.findFirst({
    where: { id: collectionId, userId },
  });
  if (!collection) throw new NotFoundError('collection not found');

  // Check name uniqueness on rename
  if (data.name && data.name !== collection.name) {
    const exists = await prisma.recipeCollection.findFirst({
      where: { userId, name: data.name, id: { not: collectionId } },
    });
    if (exists) throw new ValidationError('collection with this name already exists');
  }

  const updateData: any = {};
  if (data.name !== undefined) updateData.name = data.name;
  if (data.description !== undefined) updateData.description = data.description;

  const updated = await prisma.recipeCollection.update({
    where: { id: collectionId },
    data: updateData,
  });

  return collectionToApi(updated);
}

export async function deleteCollection(collectionId: string, userId: string) {
  const collection = await prisma.recipeCollection.findFirst({
    where: { id: collectionId, userId },
  });
  if (!collection) throw new NotFoundError('collection not found');

  await prisma.recipeCollection.delete({ where: { id: collectionId } });
}

export async function addRecipe(collectionId: string, recipeId: string, userId: string) {
  const collection = await prisma.recipeCollection.findFirst({ where: { id: collectionId, userId } });
  if (!collection) throw new NotFoundError('collection not found');

  const recipe = await prisma.recipe.findFirst({ where: { id: recipeId, userId } });
  if (!recipe) throw new NotFoundError('recipe not found');

  try {
    await prisma.recipeCollectionMember.create({
      data: { collectionId, recipeId },
    });
  } catch (err: any) {
    if (err.code !== 'P2002') throw err;
    // Already in collection, no-op
  }
}

export async function removeRecipe(collectionId: string, recipeId: string, userId: string) {
  const collection = await prisma.recipeCollection.findFirst({ where: { id: collectionId, userId } });
  if (!collection) throw new NotFoundError('collection not found');

  const recipe = await prisma.recipe.findFirst({ where: { id: recipeId, userId } });
  if (!recipe) throw new NotFoundError('recipe not found');

  await prisma.recipeCollectionMember.deleteMany({ where: { collectionId, recipeId } });
}
